# Machine Learning- Train Journey Time Prediction System

## Level 1: Data Understanding

### Objective
To understand the structure, size, quality, and characteristics of the train journey dataset before performing data cleaning and machine learning.

In [1]:
import pandas as pd
import numpy as np

In [3]:
df = pd.read_csv("Dataset1.csv")
df

,SN,Train_No,Station_Code,1A,2A,3A,SL,Station_Name,Route_Number,Arrival_time,Departure_Time,Distance
0,1,107,SWV,100,100,100,100,SAWANTWADI R,1,00:00:00,10:25:00,0
1,2,107,THVM,260,228,196,164,THIVIM,1,11:06:00,11:08:00,32
2,3,107,KRMI,345,296,247,198,KARMALI,1,11:28:00,11:30:00,49
3,4,107,MAO,490,412,334,256,MADGOAN JN.,1,12:10:00,00:00:00,78
4,1,108,MAO,100,100,100,100,MADGOAN JN.,1,00:00:00,20:30:00,0
...,...,...,...,...,...,...,...,...,...,...,...,...
186069,1,22439,NDLS,100,100,100,100,NEW DELHI,1,00:00:00,06:00:00,0
186070,2,22439,UMB,1095,896,697,1095,AMBALA CANT JN,1,08:08:00,08:10:00,199
186071,3,22439,LDH,1660,1348,1036,1660,LUDHIANA JN,1,09:19:00,09:21:00,312
186072,4,22439,JAT,2985,2408,1831,2985,JAMMU TAWI,1,12:38:00,12:40:00,577


In [26]:
df.head()

,SN,Train_No,Station_Code,1A,2A,3A,SL,Station_Name,Route_Number,Arrival_time,Departure_Time,Distance
0,1,107,SWV,100,100,100,100,SAWANTWADI R,1,00:00:00,10:25:00,0
1,2,107,THVM,260,228,196,164,THIVIM,1,11:06:00,11:08:00,32
2,3,107,KRMI,345,296,247,198,KARMALI,1,11:28:00,11:30:00,49
3,4,107,MAO,490,412,334,256,MADGOAN JN.,1,12:10:00,00:00:00,78
4,1,108,MAO,100,100,100,100,MADGOAN JN.,1,00:00:00,20:30:00,0


## Task 1.1: Dataset Size and Structure

In [5]:
print("Total Records:", df.shape[0])
print("Total Columns:", df.shape[1])

Total Records: 186074
Total Columns: 12


In [6]:
print("Column Names:")
print(df.columns.tolist())

Column Names:
['SN', 'Train_No', 'Station_Code', '1A', '2A', '3A', 'SL', 'Station_Name', 'Route_Number', 'Arrival_time', 'Departure_Time', 'Distance']


## Dataset Information

In [27]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 186074 entries, 0 to 186073
Data columns (total 12 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   SN              186074 non-null  int64 
 1   Train_No        186074 non-null  int64 
 2   Station_Code    186074 non-null  object
 3   1A              186074 non-null  int64 
 4   2A              186074 non-null  int64 
 5   3A              186074 non-null  int64 
 6   SL              186074 non-null  int64 
 7   Station_Name    186074 non-null  object
 8   Route_Number    186074 non-null  int64 
 9   Arrival_time    186074 non-null  object
 10  Departure_Time  186074 non-null  object
 11  Distance        186074 non-null  int64 
dtypes: int64(8), object(4)
memory usage: 17.0+ MB


In [28]:
df.dtypes

SN                 int64
Train_No           int64
Station_Code      object
1A                 int64
2A                 int64
3A                 int64
SL                 int64
Station_Name      object
Route_Number       int64
Arrival_time      object
Departure_Time    object
Distance           int64
dtype: object

## Task 1.2: Train-wise Starting and Ending Stations

In [9]:
df[["Train_No", "Route_Number", "Station_Name"]].head(20)

,Train_No,Route_Number,Station_Name
0,107,1,SAWANTWADI R
1,107,1,THIVIM
2,107,1,KARMALI
3,107,1,MADGOAN JN.
4,108,1,MADGOAN JN.
5,108,1,KARMALI
6,108,1,THIVIM
7,108,1,SAWANTWADI R
8,128,1,MADGOAN JN.
9,128,1,KARMALI


In [11]:
df_sorted = df.sort_values(["Train_No", "Route_Number"])
df_sorted 

,SN,Train_No,Station_Code,1A,2A,3A,SL,Station_Name,Route_Number,Arrival_time,Departure_Time,Distance
0,1,107,SWV,100,100,100,100,SAWANTWADI R,1,00:00:00,10:25:00,0
1,2,107,THVM,260,228,196,164,THIVIM,1,11:06:00,11:08:00,32
2,3,107,KRMI,345,296,247,198,KARMALI,1,11:28:00,11:30:00,49
3,4,107,MAO,490,412,334,256,MADGOAN JN.,1,12:10:00,00:00:00,78
4,1,108,MAO,100,100,100,100,MADGOAN JN.,1,00:00:00,20:30:00,0
...,...,...,...,...,...,...,...,...,...,...,...,...
186064,8,99908,AKRD,195,176,157,138,AKURDI,1,23:30:00,23:31:00,19
186065,9,99908,DEHR,220,196,172,148,DEHU ROAD,1,23:35:00,23:36:00,24
186066,10,99908,BGWI,240,212,184,156,BEGDAEWAI,1,23:39:00,23:40:00,28
186067,11,99908,GRWD,255,224,193,162,GHORAWADI,1,23:41:00,23:42:00,31


In [13]:
train_summary = df_sorted.groupby("Train_No").agg(
    Start_Station=("Station_Name", "first"),
    End_Station=("Station_Name", "last")
).reset_index()
train_summary

,Train_No,Start_Station,End_Station
0,107,SAWANTWADI R,MADGOAN JN.
1,108,MADGOAN JN.,SAWANTWADI R
2,128,MADGOAN JN.,CHHATRAPATI
3,290,DELHI-SAFDAR,DELHI-SAFDAR
4,401,AURANGABAD,VARANASI JN.
...,...,...,...
11108,99904,PUNE JN.,TALEGAON
11109,99905,TALEGAON,SHIVAJINAGAR
11110,99906,PUNE JN.,TALEGAON
11111,99907,TALEGAON,PUNE JN.


In [30]:
train_summary.head(20)

,Train_No,Start_Station,End_Station
0,107,SAWANTWADI R,MADGOAN JN.
1,108,MADGOAN JN.,SAWANTWADI R
2,128,MADGOAN JN.,CHHATRAPATI
3,290,DELHI-SAFDAR,DELHI-SAFDAR
4,401,AURANGABAD,VARANASI JN.
5,421,LUCKNOW JN.,SHRI MATA VA
6,422,SHRI MATA VA,LUCKNOW JN.
7,477,SIRSA,SIRSA
8,502,RAJENDRANAGA,AMBALA CANTT
9,504,PATNA JN.,BATHINDA JN


## Task 1.3: Distance and Number of Stops Statistics

In [31]:
df["Distance"].describe()

count    186074.000000
mean        281.353838
std         483.743964
min           0.000000
25%          23.000000
50%          73.000000
75%         291.000000
max        4260.000000
Name: Distance, dtype: float64

In [16]:
print("Minimum Distance:", df["Distance"].min())
print("Maximum Distance:", df["Distance"].max())
print("Average Distance:", df["Distance"].mean())
print("Median Distance:", df["Distance"].median())

Minimum Distance: 0
Maximum Distance: 4260
Average Distance: 281.3538377204768
Median Distance: 73.0


In [ ]:
stops_per_train = df_sorted.groupby("Train_No").size().reset_index(
    name="Number_of_Stops"
)

In [19]:
stops_per_train.head(20)
stops_per_train["Number_of_Stops"].describe()

count    11113.000000
mean        16.743814
std         12.993123
min          2.000000
25%          8.000000
50%         15.000000
75%         22.000000
max        118.000000
Name: Number_of_Stops, dtype: float64

## Task 1.4: Data Quality Assessment

In [20]:
missing_values = pd.DataFrame({
    "Missing_Count": df.isnull().sum(),
    "Missing_Percentage": df.isnull().mean() * 100
})

missing_values

,Missing_Count,Missing_Percentage
SN,0,0.0
Train_No,0,0.0
Station_Code,0,0.0
1A,0,0.0
2A,0,0.0
3A,0,0.0
SL,0,0.0
Station_Name,0,0.0
Route_Number,0,0.0
Arrival_time,0,0.0


In [21]:
duplicate_count = df.duplicated().sum()

print("Total Duplicate Records:", duplicate_count)

Total Duplicate Records: 0


In [22]:
df[df["Distance"] < 0]

,SN,Train_No,Station_Code,1A,2A,3A,SL,Station_Name,Route_Number,Arrival_time,Departure_Time,Distance


In [23]:
df[df["Train_No"] <= 0]

,SN,Train_No,Station_Code,1A,2A,3A,SL,Station_Name,Route_Number,Arrival_time,Departure_Time,Distance


In [24]:
df[df["Route_Number"] < 0]

,SN,Train_No,Station_Code,1A,2A,3A,SL,Station_Name,Route_Number,Arrival_time,Departure_Time,Distance


In [25]:
print("Arrival Time Examples:")
print(df["Arrival_time"].head(20).tolist())

print("\nDeparture Time Examples:")
print(df["Departure_Time"].head(20).tolist())

Arrival Time Examples:
['00:00:00', '11:06:00', '11:28:00', '12:10:00', '00:00:00', '21:04:00', '21:26:00', '22:25:00', '19:40:00', '20:18:00', '20:40:00', '21:16:00', '21:38:00', '21:54:00', '22:18:00', '22:40:00', '22:56:00', '23:52:00', '00:34:00', '01:10:00']

Departure Time Examples:
['10:25:00', '11:08:00', '11:30:00', '00:00:00', '20:30:00', '21:06:00', '21:28:00', '00:00:00', '19:40:00', '20:20:00', '20:42:00', '21:18:00', '21:40:00', '21:56:00', '22:20:00', '22:42:00', '22:58:00', '23:57:00', '00:36:00', '01:12:00']


## Level 1 Summary

The dataset was initially inspected to understand its size, structure, train-wise station information, distance characteristics, number of stops, missing values, duplicate records, and potentially incorrect values.

The dataset contains 186,074 records and 12 columns.

The findings from this analysis will be used in Level 2 for data cleaning and feature engineering.